wasm_runtime/src/c_api.rs

```diff
@@ -3,254 +3,398 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
-//! c_api.rs
+//! `c_api.rs`
 //!
 //! This file contains the API functions for the C language
 
-use std::ffi::{c_char, c_uchar};
-use std::slice;
+use std::ffi::{c_int, c_char, c_uchar};
 
-use crate::config::WASM_RUNTIME_CONFIG;
 use crate::ffi_utils::*;
-use crate::wasm_engine::{init_module, run_module};
+use crate::module::WasmModule;
+use crate::config::WasmConfig;
+use crate::execution_ctx::WasmExecutionCtx;
 
-/// Set the root directory for loading Wasm modules.
+
+/// Load a Wasm Module from disk.
+///
+/// All successfully loaded Wasm modules are stored in a `HashMap`.
+/// This implies that:
+///  - The `path` (also used as module's id) must point to an existing file.
+///  - The file pointed by `path` must be a valid .wasm module.
 ///
+/// In case of error, the reason is printed to stderr and returns -1.
+/// Otherwise, it returns 0.
+/// 
 /// Due to String management differences between C and Rust, this function uses `unsafe {}` code.
 /// So `path` must be a valid pointer to a null-terminated C char array. Otherwise, code might panic.
-///
 /// In addition, `path` must contain valid ASCII chars that can be converted into UTF-8 encoding.
-/// Otherwise, the root directory will be an empty string.
 ///
 /// # Examples (C Code)
 ///
 /// ```
-/// wasm_config_set_root("/var/www/wasm");
+/// wasm_module_load("/var/www/wasm/python3_11.wasm");
+/// wasm_module_load("/var/www/wasm/php8.wasm");
 /// ```
 #[no_mangle]
-pub extern "C" fn wasm_config_set_root(path: *const c_char) {
+pub extern "C" fn wasm_module_load(path: *const c_char) -> c_int {
     let path_str = const_c_char_to_str(path);
 
-    let mut config = WASM_RUNTIME_CONFIG.write()
-        .expect("ERROR! Poisoned RwLock WASM_RUNTIME_CONFIG on write()");
-
-    config.path.clear();
-    config.path.push(path_str);
+    match WasmModule::load_from_file(path_str) {
+        Ok(_) => 0,
+        Err(e) => {
+            eprintln!("ERROR! C-API: Couldn't load Wasm module at \"{}\": {}", path_str, e);
+            -1
+        }
+    }
 }
 
-/// Set the Wasm module filename
+
+/// Creates a new Wasm Config given an identifier.
+/// The identifier must be unique.
 ///
+/// In case of error, the reason is printed to stderr and returns -1.
+/// Otherwise, it returns 0.
+/// 
 /// Due to String management differences between C and Rust, this function uses `unsafe {}` code.
-/// So `filename` must be a valid pointer to a null-terminated C char array. Otherwise, code might panic.
-///
-/// In addition, `filename` must contain valid ASCII chars that can be converted into UTF-8 encoding.
-/// Otherwise, the root directory will be an empty string.
+/// So `config_id` must be a valid pointer to a null-terminated C char array. Otherwise, code might panic.
```

Comment:
I do wonder in general if `config_id` and `module_id` should be internally a null-terminated string, or if an unsigned integer would be more appropriate for the task. We could intern the string, and in our code, keep integers to refer to config and module id's. Just an idea, this can be improved later I'd say.

Reply:
Yes, that could be an option we can explore.
The reason why they are converted into strings is that is easier later to use them as the Key in a `HashMap`.